### PR TITLE
Adding the ability to support large file uploads with intermediate saving checkpoints

### PR DESCRIPTION
**Pull request by ivanleomk**

### CODE REVIEW

File: applications/wikipedia/main.py [hunk 1]
```diff
@@ -219,17 +231,26 @@ def embed_dataset(down_scale: float = 0.005, batch_size: int = 512 * 50):
...
-        pq.write_table(table, dataset_file)
-        dataset = load_dataset("parquet", data_files=dataset_file)
-        dataset.push_to_hub(dataset_name, token=os.environ["HUGGINGFACE_TOKEN"])
+        checkpoint_dir = f"{cache_dir}/checkpoints"
+        path = f"{checkpoint_dir}/{dataset_file}"
+        print(f"Saving to disk at {path}")
+        os.makedirs(checkpoint_dir,exist_ok=True)
+        pq.write_table(table, path)
+        del dataset
+        for _ in range(3):
```
Comment (jxnl) [inline applications/wikipedia/main.py]: Do we still need this

File: applications/wikipedia/upload.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    api = HfApi(token=os.environ["HUGGINGFACE_TOKEN"])
+    api.create_repo(repo_id=dataset_name, private=False, repo_type="dataset",exist_ok=True)
+
+
+    print(f"Pushing to hub {dataset_name}")
+    # dataset = load_dataset(f"{cache_dir}/wikipedia")
+    start = time.perf_counter()
+    # dataset.push_to_hub(dataset_name)
+    api.upload_folder(
+        folder_path=f"{cache_dir}/wikipedia",
```
Comment (jxnl) [inline applications/wikipedia/upload.py]: Can you update this to the checkpoints and then try a 25% batch and see how long it takes.

**Review by jxnl — CHANGES_REQUESTED** [review]

File: applications/wikipedia/main.py (hunk 1 above)
Comment (ivanleomk) [reply applications/wikipedia/main.py]: Have removed it in the latest commit. I'm just working with a separate volume now. `del dataset` seems to not work, not too sure why. I tried deleting the variable immediately after generating the batches but it still doesn't release the lock on the file

File: applications/wikipedia/upload.py (hunk 2 above)
Comment (ivanleomk) [reply applications/wikipedia/upload.py]:
```
  step b1a6b6e1eff0e9e0fa2a6d579b8f38b93af9bec1e499c724818f11cc2186b2c0 completed (still 1 to go).
data-00124-of-00127.arrow: 512MB [00:01, 512MB/s]                           
data-00125-of-00127.arrow: 512MB [00:10, 51.1MB/s]                          
data-00126-of-00127.arrow: 512MB [00:01, 401MB/s]                           
100%|██████████| 3/3 [00:12<00:00,  4.27s/it]
  step c9c3d34719f6b51dac2a4c29e9b3b20a440bb1f40a7fa669e18185dbb1aacf9e completed (still 0 to go).
All commits have been pushed.
PR is now open for reviews.
PR has been automatically merged (`merge_pr=True` was passed).
Uploaded in 758.460917796s
```

It took about 12.5 minutes to upload the entire dataset. I've found that upload time doesn't scale as much as the absolute size of the dataset though, 10% of dataset took 10 minutes while 25% took 12 minutes. We might be able to stay < 30 mins for uploading the entire dataset of embeddings ( optimistically )

File: applications/wikipedia/download.py [hunk 3]
```diff
@@ -51,9 +51,15 @@ def check_dataset_exists():
     print(dataset)  # Print out a quick summary of the dataset
     print("Dataset loaded.")
 
+    # Now try to save to disk
+    with open(f"{cache_dir}/new_text_file.txt", "w") as file:
```
Comment (jxnl) [inline applications/wikipedia/download.py]: lets clean this up

File: applications/wikipedia/main.py [hunk 4]
```diff
@@ -20,13 +20,18 @@
...
+checkpoint_dir = f"{checkpoint_volume_path}/{dataset_name}"
 data_dir = f"{cache_dir}/{dataset_name}"
 DATA_PATH = Path(data_dir)
 
-PUSH_TO_HUB = False
-dataset_name = f"567-labs/wikipedia-embedding-{MODEL_SLUG}-sample"
-dataset_file = "wiki-embeddings.parquet"
+SAVE_TO_DISK = True
+dataset_name = f"567-labs/wikipedia-embedding-{MODEL_SLUG}-debug"
+hf_dataset_name = "567-labs/upload-25"
```
Comment (jxnl) [inline applications/wikipedia/main.py]: lets clean this up and prepare for a 100%

File: applications/wikipedia/main.py [hunk 5]
```diff
@@ -219,17 +232,38 @@ def embed_dataset(down_scale: float = 0.005, batch_size: int = 512 * 50):
...
+        checkpoint_volume.commit()
+        print(f"Saved checkpoint at {path_parent_folder}")
+
+        api = HfApi(token=os.environ["HUGGINGFACE_TOKEN"])
+        api.create_repo(repo_id=hf_dataset_name, private=False, repo_type="dataset",exist_ok=True)
+
+
+        print(f"Pushing to hub {hf_dataset_name}")
+        start = time.perf_counter()
+        api.upload_folder(
```
Comment (jxnl) [inline applications/wikipedia/main.py]: does this mean we don't need upload,py?

**Comment by ivanleomk** [comment]
Made the following changes

- deleted `local.py` and `upload.py` since I don't think they're needed now
- refactored out some of the logic into helper functions so that it reads a bit more easily
- Cleaned up the vars in `main.py` so that config and constants variables are all capitalised by default 
- cleaned up `download.py` and removed the useless functions.

Also added an example of how to use the Modal cli `volume` feature with an example for `ls` and `rm` inside the Readme file.

**Pull request merged**